libunwind/src/UnwindRegistersRestore.S

```diff
@@ -1169,7 +1169,11 @@ DEFINE_LIBUNWIND_FUNCTION(_ZN9libunwind15Registers_riscv6jumptoEv)
     ILOAD x\i, (RISCV_ISIZE * \i)(a0)
   .endr
   // skip a0 for now
+#if defined(__riscv_abi_rve)
```

Comment:
Should this be based on the ABI or on the E extension? It's possible to use the ILP32E/ILP64E on hardware that has all 32 registers.

Reply:
I deliberately chose to base it on the ABI so that, in the "ilp32e on hardware with 32 registers" situation, code won't try to touch the extra registers. This would allow "evil" code to freely use them to stash all kinds of extra state and depend on the fact that compiled code will leave it alone.

Reply:
I don't understand how the non-e version of this list relates to either the ABI or the architecture, so it's really hard to know what this should be doing in the e case. I'm also somewhat surprised that the list in this function is different from the one below, in save function. Shouldn't these be somewhat symmetric (barring having to do things with a few registers which represent unwinder state, like in this case `a0`).

The non-e version is restoring all the argument registers, some (not all) of the callee-saves, and some (not all) of the temporary registers. Are there good docs as to why it is doing so?

I think we might need to restore according to the architecture rather than the ABI, so that we restore live temporary registers correctly, but I'm not entirely sure, and if i could understand what the non-e case was doing better, I'd be able to give a clearer steer as to which this should be based on.

Reply:
RV32E is an alternative "base integer instruction set" for RISC-V, in contrast to the "normal" RV32I/RV64I. The difference is that it only has 16 registers instead of 32 registers. This allows very small embedded CPU cores to be made even smaller. The instruction encoding is otherwise the same, so RV32E code is a subset of RV32I code.

ILP32E is an ABI/calling convention designed to be used with the RV32E ISA. Again, the only difference is that half the registers are gone. However, because RV32E is a subset of RV32I, code built for this ABI can run on cores with either 16 or 32 registers. [This doc](https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-cc.adoc) says "If used with an ISA that has any of the registers x16-x31 and f0-f31, then these registers are considered temporaries."

The non-e case seems to be saving/restoring all registers (just like every other architecture)? I'm not sure where you're seeing that it doesn't save all of them?

The reason the lists aren't symmetric is because register `a0` is `x10`, so restoring it is deferred until the end. I'm not sure why it doesn't just use one list, but this pattern is also used for MIPS and LoongArch.

Reply:
In general, libunwind seems to be somewhat inconsistent with when it checks ISA vs ABI defines?

* i386 doesn't touch any FPU registers at all. not sure what happens if `vectorcall` gets involved?
* x86_64 only restores XMM registers for `_WIN64` which is an ABI-related define
* PPC32/PPC64 checks `__VSX__`/`__NO_FPRS__`/`__ALTIVEC__` which are ISA-related defines
* ARM has its own special unique logic for handling FPU registers involving runtime flags
* MIPS checks `__mips_hard_float` which is an ABI-related define but also `__mips_isa_rev` which is an ISA-related define
* SPARC has separate functions for two different ABIs
* RISC-V checks `__riscv_flen` which is an ISA-related define
* LoongArch checks `__loongarch_frlen` which is an ISA-related define
* aarch64/or1k/hexagon/s390x do not have any variations to handle

Reply:
> The non-e case seems to be saving/restoring all registers (just like every other architecture)? I'm not sure where you're seeing that it doesn't save all of them?

Sorry, my fault, I didn't realise there were two loops in this function and thus two different lists of registers.

> 
> The reason the lists aren't symmetric is because register a0 is x10, so restoring it is deferred until the end. I'm not sure why it doesn't just use one list, but this pattern is also used for MIPS and LoongArch.

Yeah, I'm happy with this.

Broadly, going back to your/craig's original question: I feel this should be keyed on architecture, not ABI. This function is restoring all registers, including temporaries, not just caller-saves, so on the e-abi it should be restoring all temporaries as well. (This is different to my change to the save/restore routines, which only touch callee-saves).

Reply:
Okay, I apparently did not realize that you can explicitly force `-march=rv32i -mabi=ilp32e` in which case clang will indeed start using x16-x31 as temporaries. This means that `__riscv_32e` is probably the correct define to be checking. I will update the PR.

Incidentally, does this mean that MIPS soft-float-ABI-but-FPU-enabled might be buggy?

Reply:
> Okay, I apparently did not realize that you can explicitly force `-march=rv32i -mabi=ilp32e` in which case clang will indeed start using x16-x31 as temporaries. This means that `__riscv_32e` is probably the correct define to be checking. I will update the PR.

Yeah, the intention here is that you should be able to use an E-abi library on the I architecture, without modification. If you compile your app for the I architecture to get additional temporaries, then you need to make sure you use a libunwind that knows about the additional temporaries it might need to restore.

> Incidentally, does this mean that MIPS soft-float-ABI-but-FPU-enabled might be buggy?

Sounds like it could be, but I'm no MIPS expert.